internal/scheduler/reports.go

```diff
@@ -5,442 +5,394 @@ import (
+// A mutex protects against concurrent writes.
+type SchedulingContextRepository struct {
+	// Maps executor id to *SchedulingContext.
+	// The most recent attempt.
+	mostRecentSchedulingContextByExecutorP atomic.Pointer[SchedulingContextByExecutor]
+	// The most recent attempt where a non-zero amount of resources were scheduled.
+	mostRecentSuccessfulSchedulingContextByExecutorP atomic.Pointer[SchedulingContextByExecutor]
+	// Maps queue name to QueueSchedulingContextByExecutor.
+	// The most recent attempt.
+	mostRecentQueueSchedulingContextByExecutorByQueueP atomic.Pointer[map[string]QueueSchedulingContextByExecutor]
+	// The most recent attempt where a non-zero amount of resources were scheduled.
+	mostRecentSuccessfulQueueSchedulingContextByExecutorByQueueP atomic.Pointer[map[string]QueueSchedulingContextByExecutor]
+	// Maps job id to JobSchedulingContextByExecutor.
+	// We limit the number of job contexts to store to control memory usage.
+	mostRecentJobSchedulingContextByExecutorByJobId *lru.Cache
+	// Store all executor ids seen so far in a set.
+	// Used to ensure all executors are included in reports.
+	executorIds map[string]bool
+	// All executors in sorted order.
+	sortedExecutorIdsP atomic.Pointer[[]string]
+	// Protects the fields in this struct from concurrent and dirty writes.
+	mu sync.Mutex
 }
 
-func NewSchedulingReportsRepository(maxQueueSchedulingReports, maxJobSchedulingReports int) *SchedulingReportsRepository {
-	mostRecentJobSchedulingReports, err := lru.New(maxJobSchedulingReports)
-	if err != nil {
-		panic(errors.WithStack(err))
-	}
-	mostRecentQueueSchedulingReports, err := lru.New(maxQueueSchedulingReports)
+type (
+	SchedulingContextByExecutor      map[string]*SchedulingContext
+	QueueSchedulingContextByExecutor map[string]*QueueSchedulingContext
+	JobSchedulingContextByExecutor   map[string]*JobSchedulingContext
+)
+
+func NewSchedulingContextRepository(maxJobSchedulingContextsPerExecutor uint) (*SchedulingContextRepository, error) {
+	jobSchedulingContextByExecutorByJobId, err := lru.New(int(maxJobSchedulingContextsPerExecutor))
 	if err != nil {
-		panic(errors.WithStack(err))
-	}
-	return &SchedulingReportsRepository{
-		MostRecentJobSchedulingReports:   mostRecentJobSchedulingReports,
-		MostRecentQueueSchedulingReports: mostRecentQueueSchedulingReports,
+		return nil, err
 	}
+	rv := &SchedulingContextRepository{
+		mostRecentJobSchedulingContextByExecutorByJobId: jobSchedulingContextByExecutorByJobId,
+		executorIds: make(map[string]bool),
+	}
+	mostRecentSchedulingContextByExecutor := make(SchedulingContextByExecutor)
+	mostRecentSuccessfulSchedulingContextByExecutor := make(SchedulingContextByExecutor)
+	mostRecentQueueSchedulingContextByExecutorByQueue := make(map[string]QueueSchedulingContextByExecutor)
+	mostRecentSuccessfulQueueSchedulingContextByExecutorByQueue := make(map[string]QueueSchedulingContextByExecutor)
+	sortedExecutorIds := make([]string, 0)
```

Comment:
I don't think this is correct.  You need the state updae to be appplied atomically otherwise e.g. the reader may get a list of executorIds that don't correspond to the map keys.  I think what you want is a struct that contains all of the state and a single atomic reference which holds it.

Reply:
That should be safe. Only `sortedExecutorIds` is used by the readers so it's fine for it to be briefly inconsistent with the map.